File: awa/awa/sdk/utils/activity/read_directory_activity.py
from datetime import timedelta
from pathlib import Path

from temporalio import workflow

from awa.sdk import constants
from awa.sdk.models.read_directory_result import ReadDirectoryResult


async def read_directory_activity(
    source_path: str | Path,
    ignore_file_path: str | Path | None = None,
) -> list[ReadDirectoryResult]:
    """Read a directory and return detailed information about its contents.

    Args:
        source_path: The path to the directory to read. Can be a string or Path object.
        ignore_file_path: Optional path to an ignore file (like .gitignore) to exclude files.
            Can be a string or Path object. Defaults to None.

    Returns:
        list[ReadDirectoryResult]: A list of ReadDirectoryResult objects containing
            detailed information about each file and directory.

    """
    raw_results = await workflow.execute_activity(
        constants.ACTIVITY_READ_DIRECTORY,
        args=[str(source_path), str(ignore_file_path) if ignore_file_path else None],
        task_queue=constants.AWA_DEFAULT_TASK_QUEUE,
        start_to_close_timeout=timedelta(seconds=constants.FILE_IO_TIMEOUT_SECONDS * 10),
    )
    return [ReadDirectoryResult(**result) for result in raw_results]

File: awa/sdk_dist/csharp/Awa.Client/Utilities/Activity/ReadDirectoryActivity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Awa.Client.Constants;
using Awa.Client.Models;

namespace Awa.Client.Utilities.Activity
{
    public static partial class ActivityUtilities
    {
        /// <summary>
        /// Read a directory and return detailed information about its contents.
        /// </summary>
        /// <param name="sourcePath">The path to the directory to read.</param>
        /// <param name="ignoreFilePath">Optional path to an ignore file (like .gitignore) to exclude files. Defaults to null.</param>
        /// <returns>A list of ReadDirectoryResult objects containing detailed information about each file and directory.</returns>
        public static async Task<List<ReadDirectoryResult>> ReadDirectoryActivity(
            string sourcePath,
            string? ignoreFilePath = null)
        {
            var rawResults = await Temporalio.Workflows.Workflow.ExecuteActivityAsync<object[]>(
                AwaConstants.ActivityReadDirectory,
                new object?[] { sourcePath, ignoreFilePath },
                new()
                {
                    TaskQueue = AwaConstants.AwaDefaultTaskQueue,
                    StartToCloseTimeout = TimeSpan.FromSeconds(AwaConstants.FileIoTimeoutSeconds * 10)
                });

            return rawResults.Select(result =>
            {
                var dict = result as Dictionary<string, object> ?? new Dictionary<string, object>();
                return new ReadDirectoryResult
                {
                    File = dict.TryGetValue("file", out var file) ? file?.ToString() ?? "" : "",
                    Content = dict.TryGetValue("content", out var content) ? content?.ToString() ?? "" : ""
                };
            }).ToList();
        }
    }
}
